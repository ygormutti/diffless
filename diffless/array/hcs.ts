import { Equals, Weigh } from '../model';

/**
 * Heaviest Common Subsequence model and implementation
 */

/**
 * Function type that finds the HCS between two arrays
 */
export type HCS = <TItem>(
    equal: Equals<TItem>,
    weigh: Weigh<TItem>,
    left: TItem[],
    right: TItem[],
) => HCSResult<TItem>;

/**
 * The Heaviest Common Subsequence of two arrays with items from the respective arrays
 */
export class HCSResult<TItem> {
    constructor(
        readonly weight: number,
        readonly leftHCS: TItem[],
        readonly rightHCS: TItem[],
    ) { }

    get length() {
        return this.leftHCS.length;
    }
}

class HCSCell<TItem> {
    constructor(
        readonly weight: number = 0,
        readonly hcs: [TItem, TItem][] = [],
    ) { }
}

/**
 * Solves HCS with quadratic time and linear space complexity
 */
export function dynamicProgrammingHCS<TItem>(
    equals: Equals<TItem>,
    weigh: Weigh<TItem>,
    left: TItem[],
    right: TItem[],
): HCSResult<TItem> {
    if (!left || !left.length || !right || !right.length) {
        return new HCSResult(0, [], []);
    }

    const width = left.length + 1;
    const height = right.length + 1;

    let prevRowArray: HCSCell<TItem>[] = new Array(width);
    prevRowArray.fill(new HCSCell());

    for (let row = 1; row < height; row++) {
        const rowArray: HCSCell<TItem>[] = new Array(width);
        rowArray[0] = new HCSCell();

        for (let col = 1; col < width; col++) {
            const leftItem = left[col - 1];
            const rightItem = right[row - 1];
            if (equals(leftItem, rightItem)) {
                const prevCell = prevRowArray[col - 1];
                const newWeight = prevCell.weight + weigh(leftItem);
                rowArray[col] = new HCSCell(newWeight, [...prevCell.hcs, [leftItem, rightItem]]);
            } else {
                const previousCellCandidates = [rowArray[col - 1], prevRowArray[col]];
                previousCellCandidates.sort((a, b) => {
                    if (a.weight !== b.weight) {
                        return b.weight - a.weight;
                    }
                    return a.hcs.length - b.hcs.length;
                });
                rowArray[col] = previousCellCandidates[0];
            }
        }
        prevRowArray = rowArray;
    }

    const lastCell = prevRowArray[prevRowArray.length - 1];
    const { hcs, weight } = lastCell;
    const leftHCS = hcs.map(p => p[0]);
    const rightHCS = hcs.map(p => p[1]);
    return new HCSResult(weight, leftHCS, rightHCS);
}
