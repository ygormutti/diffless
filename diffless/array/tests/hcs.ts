import { dynamicProgrammingHCS } from '../hcs';

describe('dynamicProgrammingHCS', () => {
    it('should work as LCS with uniform weight', () => {
        const left = 'XMJYAUZ';
        const right = 'mzjawxu';

        const equal = (l: string, r: string) => l.toLowerCase() === r.toLowerCase();
        const weight = () => 1;
        const result = dynamicProgrammingHCS(equal, weight, left.split(''), right.split(''));

        expect(result).toMatchSnapshot();
        expect(result.leftHCS.join('')).toBe('MJAU');
    });

    it('should work with variable weight', () => {
        const left = 'XMJYAUZ';
        const right = 'mzjawxu';

        const equal = (l: string, r: string) => l.toLowerCase() === r.toLowerCase();
        const weight = (i: string) => i === 'Z' ? 10 : 1;
        const result = dynamicProgrammingHCS(equal, weight, left.split(''), right.split(''));

        expect(result).toMatchSnapshot();
        expect(result.leftHCS.join('')).toBe('MZ');
    });
});
