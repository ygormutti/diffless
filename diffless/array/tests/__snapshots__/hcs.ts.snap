// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dynamicProgrammingHCS should work as LCS with uniform weight 1`] = `
HCSResult {
  "leftHCS": Array [
    "M",
    "J",
    "A",
    "U",
  ],
  "rightHCS": Array [
    "m",
    "j",
    "a",
    "u",
  ],
}
`;

exports[`dynamicProgrammingHCS should work with variable weight 1`] = `
HCSResult {
  "leftHCS": Array [
    "M",
    "Z",
  ],
  "rightHCS": Array [
    "m",
    "z",
  ],
}
`;
